zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,195 @@
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = this.converter.convert(nodeData);
+            }
+        } catch (Exception e) {
+            throw new EndpointGroupException(
+                    "failed to connect to ZooKeeper:  " + zkConnectionStr + " with error:  " + e.getMessage());
+        }
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultZkNodeValueConverter}.
+     * @param zkConnectionStr A connection string containing a comma
+     *                          separated list of host:port pairs , each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultZkNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return endpointList;
+    }
+
+    final class ZkWatcher implements Watcher, StatCallback {
+        private int retryDelay = 1000; //start with one second
+
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(zNode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(zNode, true, this, null);
+                }
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                case NOAUTH:
+                    close();
+                    logger.error("Underlying ZooKeeper connection has been closed " +
+                                 "due to session expired or authentication error");
+                    return;
+                default:
+                    // Retry errors, by using a exponential backoff strategy
+                    if (retryDelay < MAX_RETRY_DELAY) {
+                        try {
+                            Thread.sleep(retryDelay);
+                        } catch (InterruptedException e) {
+                            throw new EndpointGroupException(e);
+                        }
+                        retryDelay = Math.min(MAX_RETRY_DELAY, retryDelay * 2);
+                    } else { //retry at maximum delay
+                        retryDelay = MAX_RETRY_DELAY;
```

Comment:
I don't think we need this `else` block and `if` statement above because we made sure `retryDelay` is always equal to or less than `MAX_RETRY_DELAY` using `Math.min(...) below.